crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -232,91 +233,204 @@ impl ConsumerTask {
-                            .split_partition_queue(&topic, partition) {
-                            Some(q) => q,
-                            None => break Err(Error::TopicPartitionSplit(topic.clone(), partition))
-                        };
-
-                        let task = topic_partition_queue_consumption_loop(
-                            self.sender.clone(),
-                            topic.clone(), partition,
-                            topic_partition_consumer,
-                            Arc::clone(&consumer),
-                            consumer_group_id.clone()
-                        );
-
-                        if let Ok(task_id) = self.task_center.spawn_child(TaskKind::Ingress, "partition-queue", None, task) {
-                            e.insert(task_id);
-                        } else {
-                            break Ok(());
+        // we have to poll the main consumer for callbacks to be processed, but we expect to only see messages on the partitioned queues
+        tokio::select! {
+            res = consumer.recv() => {
+                match res {
+                    // We shouldn't see any messages on the main consumer loop
+                    Ok(msg) => Err(Error::UnexpectedMainQueueMessage(msg.topic().into(), msg.partition())),
+                    Err(e) => Err(e.into()),
+                }
+            }
+            _ = &mut rx => {
+                 Ok(())
+            }
+        }
+    }
+}
+
+struct ConsumerDrop(Arc<MessageConsumer>);
+
+impl std::ops::Deref for ConsumerDrop {
+    type Target = Arc<MessageConsumer>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Drop for ConsumerDrop {
+    fn drop(&mut self) {
+        debug!(
+            "Stopping consumer with id {}",
+            self.context().consumer_group_id
+        );
+
+        let task_center = &self.context().task_center;
+        self
+            .context()
+            .topic_partition_tasks
+            .lock()
+            .unwrap()
+            .drain()
+            .for_each(|(partition, task_id)| {
+                debug!("Stopping partitioned consumer for partition {partition} due to the consumer stopping");
+                if let Some(handle) = task_center.cancel_task(task_id) {
+                    handle.abort()
+                }
+            });
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct TopicPartition(String, i32);
+
+impl<'a> From<TopicPartitionListElem<'a>> for TopicPartition {
+    fn from(value: TopicPartitionListElem<'a>) -> Self {
+        Self(value.topic().into(), value.partition())
+    }
+}
+
+impl Display for TopicPartition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}/{}", self.0, self.1)
+    }
+}
+
+struct RebalanceContext {
+    task_center: TaskCenter,
+    consumer: OnceLock<Weak<MessageConsumer>>,
+    topic_partition_tasks: Mutex<HashMap<TopicPartition, TaskId>>,
```

Comment:
You probably want to use parking_lot's Mutex here to release on poison since you need to lock on Drop.